File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    public int solution(String dirs) {
        
        int x = 0;
        int y = 0;
        
        HashSet<String> set = new HashSet<>();
        
        for(int i=0; i<dirs.length(); i++){
            int nowx = x, nowy = y;
            char now = dirs.charAt(i);
            switch(now){
                case 'U' : 
                    if(x < 5) x++;
                    break;
                case 'D' :
                    if(x > -5) x--;
                    break;
                case 'L' :
                    if(y > -5) y--;
                    break;
                case 'R' :
                    if(y < 5) y++;
                    break;
            }
            
            if (nowx == x && nowy == y) continue;
            String path1 = x + " " + y + "->" + nowx + " " + nowy;
            String path2 = nowx + " " + nowy + "->" + x + " " + y;
            
            set.add(path1);
            set.add(path2);
        }

        return set.size()/2;
    }
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    
    var x = 0;
    var y = 0;
    
    var path = new Map();
    
    for(let i=0; i<dirs.length; i++){
        var nowx = x;
        var nowy = y;
        var dir = dirs.charAt(i);
        if(dir == 'U' && nowx < 5){
            nowx ++;
        }else if(dir == 'D' && nowx > -5){
            nowx --;
        }else if(dir == 'L' && nowy > -5){
            nowy --;
        }else if(dir == 'R' && nowy < 5){
            nowy ++;
        }
        
        if(nowx == x && nowy == y) continue;
        
        var next1 = x + " " + y + " " + nowx + " " + nowy;
        var next2 = nowx + " " + nowy + " " + x + " " + y;
       
        path.set(next1);
        path.set(next2);
        
        x = nowx;
        y = nowy;
    }
    
    var answer = 0;
    return path.size/2;
}